File: workflow-sdk-go/runner/runner.go
// Package runner implements the poll/lock/dispatch loop for the Go SDK.
package runner

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/batnam/rochallor-engine/workflow-sdk-go/client"
	"github.com/batnam/rochallor-engine/workflow-sdk-go/handler"
	"github.com/batnam/rochallor-engine/workflow-sdk-go/metrics"
	"github.com/batnam/rochallor-engine/workflow-sdk-go/retry"
)

const (
	defaultParallelism  = 64
	defaultPollInterval = 500 * time.Millisecond
)

// Config holds runner configuration.
type Config struct {
	WorkerID     string
	Parallelism  int           // goroutines (default 64)
	PollInterval time.Duration // between poll rounds (default 500ms)
}

// Runner polls the Engine and dispatches jobs to registered handlers.
type Runner struct {
	cfg      Config
	engine   client.EngineClient
	registry *handler.Registry
}

// New creates a Runner. cfg.WorkerID must not be empty.
func New(cfg Config, engine client.EngineClient, registry *handler.Registry) *Runner {
	if cfg.Parallelism <= 0 {
		cfg.Parallelism = defaultParallelism
	}
	if cfg.PollInterval <= 0 {
		cfg.PollInterval = defaultPollInterval
	}
	return &Runner{cfg: cfg, engine: engine, registry: registry}
}

// Run starts the poll loop and blocks until ctx is cancelled.
// All in-flight handlers are drained before Run returns.
func (r *Runner) Run(ctx context.Context) {
	sem := make(chan struct{}, r.cfg.Parallelism)
	var wg sync.WaitGroup

	ticker := time.NewTicker(r.cfg.PollInterval)
	defer ticker.Stop()

	jobTypes := r.registry.JobTypes()
	slog.Info("runner: starting", "worker_id", r.cfg.WorkerID, "job_types", jobTypes, "parallelism", r.cfg.Parallelism)

	for {
		select {
		case <-ctx.Done():
			slog.Info("runner: received stop signal, draining in-flight jobs")
			wg.Wait()
			slog.Info("runner: stopped")
			return
		case <-ticker.C:
			r.pollAndDispatch(ctx, sem, &wg, jobTypes)
		}
	}
}

func (r *Runner) pollAndDispatch(ctx context.Context, sem chan struct{}, wg *sync.WaitGroup, jobTypes []string) {
	start := time.Now()
	jobs, err := r.engine.PollJobs(ctx, client.PollJobsRequest{
		WorkerID: r.cfg.WorkerID,
		JobTypes: jobTypes,
		MaxJobs:  r.cfg.Parallelism,
	})
	metrics.PollLatency.Observe(time.Since(start).Seconds())

	if err != nil {
		slog.Error("runner: poll error", "err", err)
		return
	}
	if len(jobs) == 0 {
		metrics.LockConflicts.Add(1)
		return
	}

	for _, j := range jobs {
		j := j // capture
		sem <- struct{}{}
		wg.Add(1)
		go func() {
			defer wg.Done()
			defer func() { <-sem }()
			r.dispatch(ctx, j)
		}()
	}
}

func (r *Runner) dispatch(ctx context.Context, j client.Job) {
	h, err := r.registry.Get(j.JobType)
	if err != nil {
		slog.Error("runner: no handler", "job_id", j.ID, "job_type", j.JobType)
		_ = r.engine.FailJob(ctx, j.ID, r.cfg.WorkerID, "no handler registered for "+j.JobType, false)
		return
	}

	var vars map[string]any
	if len(j.Variables) > 0 {
		_ = json.Unmarshal(j.Variables, &vars)
	}

	jctx := handler.JobContext{
		JobID:            j.ID,
		InstanceID:       j.InstanceID,
		StepID:           j.StepExecutionID,
		JobType:          j.JobType,
		RetriesRemaining: j.RetriesRemaining,
		Variables:        vars,
	}

	start := time.Now()
	slog.Info("runner: dispatching job", "job_id", j.ID, "job_type", j.JobType)
	result, err := safeCall(ctx, h, jctx)
	metrics.HandlerLatency.WithLabelValues(j.JobType).Observe(time.Since(start).Seconds())

	if err == nil {
		slog.Info("runner: job completed", "job_id", j.ID)
		metrics.JobsCompletedTotal.WithLabelValues(j.JobType, "success").Inc()
		if completeErr := r.engine.CompleteJob(ctx, j.ID, r.cfg.WorkerID, result.VariablesToSet); completeErr != nil {
			slog.Error("runner: CompleteJob failed", "job_id", j.ID, "err", completeErr)
		}
		return
	}

	// Handler failed
	isNonRetryable := retry.IsNonRetryable(err)
	if isNonRetryable {
		slog.Warn("runner: job failed (non-retryable)", "job_id", j.ID, "err", err)
	} else {
		slog.Warn("runner: job failed (retryable)", "job_id", j.ID, "err", err)
	}
	metrics.JobsCompletedTotal.WithLabelValues(j.JobType, "failure").Inc()
	if !isNonRetryable {
		metrics.RetriesTotal.WithLabelValues(j.JobType).Inc()
	}

	if failErr := r.engine.FailJob(ctx, j.ID, r.cfg.WorkerID, err.Error(), !isNonRetryable); failErr != nil {
		slog.Error("runner: FailJob failed", "job_id", j.ID, "err", failErr)
	}
}

// safeCall executes h, recovering from panics.
func safeCall(ctx context.Context, h handler.Handler, jctx handler.JobContext) (res handler.Result, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = &retry.NonRetryable{Cause: fmt.Errorf("handler panic: %v", r)}
		}
	}()
	return h(ctx, jctx)
}

File: workflow-sdk-python/src/workflow_sdk/runner/runner.py
"""Worker runner — poll / dispatch / complete loop.

Mirrors the Go SDK's goroutine-based runner using Python's ThreadPoolExecutor
for bounded concurrency.
"""

from __future__ import annotations

import logging
import signal
import threading
import time
from concurrent.futures import Future, ThreadPoolExecutor
from typing import Any

from workflow_sdk.client.interface import EngineClient
from workflow_sdk.errors import NonRetryableError, WorkflowSDKError
from workflow_sdk.handler.registry import HandlerRegistry
from workflow_sdk.metrics.metrics import Metrics
from prometheus_client import CollectorRegistry

logger = logging.getLogger(__name__)

_DEFAULT_PARALLELISM = 64
_DEFAULT_POLL_INTERVAL = 0.5  # seconds


class Runner:
    """Polls the workflow engine and dispatches jobs to registered handlers.

    The runner blocks inside :meth:`run` until ``stop_event`` is set or a
    ``SIGTERM`` / ``SIGINT`` signal is received.  All in-flight job executions
    are drained before the method returns.

    Example::

        from workflow_sdk.client import RestEngineClient
        from workflow_sdk.handler import HandlerRegistry
        from workflow_sdk.runner import Runner

        client = RestEngineClient("http://localhost:8080")
        registry = HandlerRegistry()
        registry.register("my-job", lambda ctx: {"done": True})

        runner = Runner(client, registry, worker_id="worker-1")
        runner.run()  # blocks until SIGTERM/SIGINT

    Args:
        client: Engine client (REST or future gRPC).
        registry: Handler registry mapping job types to callables.
        worker_id: Unique identifier for this worker process.
        parallelism: Maximum concurrent job executions (default 64).
        poll_interval: Seconds between poll rounds (default 0.5).
        metrics: Prometheus metrics instance. A fresh isolated registry is
                 created automatically if not provided.
    """

    def __init__(
        self,
        client: EngineClient,
        registry: HandlerRegistry,
        worker_id: str,
        parallelism: int = _DEFAULT_PARALLELISM,
        poll_interval: float = _DEFAULT_POLL_INTERVAL,
        metrics: Metrics | None = None,
    ) -> None:
        if not worker_id:
            raise ValueError("worker_id must not be empty")

        self._client = client
        self._registry = registry
        self._worker_id = worker_id
        self._parallelism = parallelism if parallelism > 0 else _DEFAULT_PARALLELISM
        self._poll_interval = poll_interval if poll_interval > 0 else _DEFAULT_POLL_INTERVAL
        self._metrics = metrics or Metrics(registry=CollectorRegistry())

    # ------------------------------------------------------------------ #
    # Public API                                                           #
    # ------------------------------------------------------------------ #

    def run(self, stop_event: threading.Event | None = None) -> None:
        """Start the polling loop and block until signalled to stop.

        Registers ``SIGTERM`` and ``SIGINT`` handlers that set the stop event.
        If ``stop_event`` is ``None``, an internal one is created.

        Args:
            stop_event: External event to signal shutdown.  Set it from
                another thread to trigger graceful shutdown.
        """
        if stop_event is None:
            stop_event = threading.Event()

        self._install_signal_handlers(stop_event)

        job_types = self._registry.job_types()
        logger.info(
            "runner: starting worker_id=%s job_types=%s parallelism=%d",
            self._worker_id,
            job_types,
            self._parallelism,
        )

        in_flight: list[Future[None]] = []

        with ThreadPoolExecutor(max_workers=self._parallelism) as executor:
            while not stop_event.is_set():
                self._poll_and_dispatch(executor, in_flight, job_types)
                # Prune completed futures to avoid memory growth
                in_flight[:] = [f for f in in_flight if not f.done()]
                stop_event.wait(timeout=self._poll_interval)

            # Drain remaining in-flight jobs
            logger.info("runner: draining %d in-flight jobs", len(in_flight))
            for future in in_flight:
                try:
                    future.result(timeout=300)
                except Exception:
                    pass  # errors were already handled inside _dispatch

        logger.info("runner: stopped")

    # ------------------------------------------------------------------ #
    # Internal methods                                                     #
    # ------------------------------------------------------------------ #

    def _poll_and_dispatch(
        self,
        executor: ThreadPoolExecutor,
        in_flight: list[Future[None]],
        job_types: list[str],
    ) -> None:
        start = time.monotonic()
        try:
            jobs = self._client.poll_jobs(
                self._worker_id, job_types, max_jobs=self._parallelism
            )
        except WorkflowSDKError as exc:
            logger.warning("runner: poll error — %s", exc)
            return
        except Exception as exc:
            logger.warning("runner: unexpected poll error — %s", exc)
            return
        finally:
            self._metrics.poll_latency.observe(time.monotonic() - start)

        if not jobs:
            self._metrics.lock_conflicts.inc()
            return

        for job in jobs:
            logger.info("runner: dispatching job_id=%s job_type=%s", job["id"], job["jobType"])
            future = executor.submit(self._dispatch, job)
            in_flight.append(future)

    def _dispatch(self, job: dict[str, Any]) -> None:
        job_id: str = job["id"]
        job_type: str = job["jobType"]

        handler = self._registry.get(job_type)
        if handler is None:
            logger.error("runner: no handler for job_type=%s job_id=%s", job_type, job_id)
            self._safe_fail(job_id, f"no handler registered for {job_type!r}", retryable=False)
            return

        ctx: dict[str, Any] = {
            "id": job_id,
            "jobType": job_type,
            "instanceId": job.get("instanceId", ""),
            "stepId": job.get("stepId", ""),
            "stepExecutionId": job.get("stepExecutionId", ""),
            "retriesRemaining": job.get("retriesRemaining", 0),
            "variables": job.get("variables") or {},
        }

        start = time.monotonic()
        try:
            result = self._safe_call(handler, ctx)
            logger.info("runner: job completed job_id=%s", job_id)
        except NonRetryableError as exc:
            self._metrics.handler_latency.labels(job_type=job_type).observe(
                time.monotonic() - start
            )
            self._metrics.jobs_completed.labels(job_type=job_type, outcome="failure").inc()
            logger.warning("runner: job failed (non-retryable) job_id=%s: %s", job_id, exc)
            self._safe_fail(job_id, str(exc), retryable=False)
            return
        except Exception as exc:
            self._metrics.handler_latency.labels(job_type=job_type).observe(
                time.monotonic() - start
            )
            self._metrics.jobs_completed.labels(job_type=job_type, outcome="failure").inc()
            self._metrics.retries_total.labels(job_type=job_type).inc()
            logger.exception("runner: job failed (retryable) job_id=%s: %s", job_id, exc)
            self._safe_fail(job_id, str(exc), retryable=True)
            return

        self._metrics.handler_latency.labels(job_type=job_type).observe(
            time.monotonic() - start
        )
        self._metrics.jobs_completed.labels(job_type=job_type, outcome="success").inc()

        variables = result if isinstance(result, dict) else {}
        try:
            self._client.complete_job(job_id, self._worker_id, variables)
        except Exception as exc:
            logger.error("runner: complete_job failed job_id=%s — %s", job_id, exc)

    @staticmethod
    def _safe_call(
        handler: Any,
        ctx: dict[str, Any],
    ) -> dict[str, Any] | None:
        """Execute handler, converting panics (unhandled exceptions) to NonRetryableError."""
        try:
            return handler(ctx)
        except NonRetryableError:
            raise
        except Exception:
            raise

    def _safe_fail(self, job_id: str, error_message: str, retryable: bool) -> None:
        try:
            self._client.fail_job(job_id, self._worker_id, error_message, retryable)
        except Exception as exc:
            logger.error("runner: fail_job error job_id=%s — %s", job_id, exc)

    @staticmethod
    def _install_signal_handlers(stop_event: threading.Event) -> None:
        def _handle(signum: int, frame: object) -> None:
            logger.info("runner: received signal %d, shutting down", signum)
            stop_event.set()

        try:
            signal.signal(signal.SIGTERM, _handle)
            signal.signal(signal.SIGINT, _handle)
        except (OSError, ValueError):
            # Can't install signal handlers in non-main threads — ignore
            pass
